Source/WebCore/page/EventHandler.cpp

```diff
@@ -4337,49 +4337,107 @@ void EventHandler::defaultBackspaceEventHandler(KeyboardEvent& event)
         event.setDefaultHandled();
 }
 
-KeyboardScrollingAnimator* EventHandler::keyboardScrollingAnimatorForFocusedNode()
+void EventHandler::stopKeyboardScrolling()
 {
-    Node* node = m_frame.document()->focusedElement();
+    auto animator = m_frame.page()->currentKeyboardScrollingAnimator();
+    if (animator)
+        animator->handleKeyUpEvent();
+}
 
-    if (!node)
-        node = m_mousePressNode.get();
+bool EventHandler::beginKeyboardScrollGesture(KeyboardScrollingAnimator* animator, KeyboardEvent& event)
+{
+    auto* platformEvent = event.underlyingPlatformEvent();
 
-    auto* scrollableArea = enclosingScrollableArea(node);
+    if (animator && platformEvent && animator->beginKeyboardScrollGesture(*platformEvent)) {
+        m_frame.page()->setCurrentKeyboardScrollingAnimator(animator);
+        return true;
+    }
+
+    return false;
+}
+
+bool EventHandler::startKeyboardScrollAnimationOnDocument(KeyboardEvent& event)
+{
+    RefPtr frame = &m_frame;
+    RefPtr view = frame->view();
+    if (!view)
+        return false;
+
+    auto* animator = view->scrollAnimator().keyboardScrollingAnimator();
+    return beginKeyboardScrollGesture(animator, event);
+}
+
+bool EventHandler::startKeyboardScrollAnimationOnRenderBoxLayer(KeyboardEvent& event, RenderBox* renderBox)
+{
+    auto* scrollableArea = renderBox->layer() ? renderBox->layer()->scrollableArea() : nullptr;
     if (!scrollableArea)
-        return nullptr;
+        return false;
 
-    return scrollableArea->scrollAnimator().keyboardScrollingAnimator();
+    auto* animator = scrollableArea->scrollAnimator().keyboardScrollingAnimator();
+    return beginKeyboardScrollGesture(animator, event);
 }
 
-void EventHandler::stopKeyboardScrolling()
+bool EventHandler::startKeyboardScrollAnimationOnRenderBoxAndItsAncestors(KeyboardEvent& event, RenderBox* renderBox)
 {
-    auto* animator = keyboardScrollingAnimatorForFocusedNode();
-    if (animator)
-        animator->handleKeyUpEvent();
+    if (startKeyboardScrollAnimationOnRenderBoxLayer(event, renderBox))
+        return true;
+
+    RenderBlock* nextRenderBlock = renderBox->containingBlock();
+
+    if (nextRenderBlock && !nextRenderBlock->isRenderView())
```

Comment:
It would be nicer to do this iteratively rather than recursively, which would involve a loop which calls `containingBlock()`.